File: src/backend/app/controllers/StandardController.java
package controllers;

import models.Standard;
import models.UseCase;
import play.db.jpa.JPAApi;
import play.libs.Json;
import play.mvc.*;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;

import javax.inject.Inject;
import jakarta.persistence.EntityManager;
import java.util.List;
import java.util.Set;
import java.util.HashSet;
import java.util.function.Function;

import actions.Secured;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@With(Secured.class)
public class StandardController extends Controller {

    private final JPAApi jpaApi;
    private static final Logger log = LoggerFactory.getLogger(StandardController.class);

    @Inject
    public StandardController(JPAApi jpaApi) {
        this.jpaApi = jpaApi;
    }

    public Result createStandard(Http.Request request) {
        JsonNode json = request.body().asJson();
        if (json == null || !json.has("name") || json.get("name").asText().trim().isEmpty()) {
            log.warn("Attempted to create standard with missing or empty name");
            return badRequest(Json.newObject().put("error", "Standard name cannot be empty."));
        }
        String name = json.get("name").asText().trim();
        String description = json.has("description") ? json.get("description").asText().trim() : "";

        List<Standard> existingStandards = jpaApi.withTransaction("default", true, (Function<EntityManager, List<Standard>>) entityManager ->
            entityManager.createQuery("SELECT s FROM Standard s WHERE LOWER(s.name) = LOWER(:name)", Standard.class)
                         .setParameter("name", name.toLowerCase())
                         .getResultList()
        );

        if (!existingStandards.isEmpty()) {
            log.warn("Attempted to create a standard that already exists: {}", name);
            return status(CONFLICT, Json.newObject().put("error", "Standard with name '" + name + "' already exists."));
        }

        Standard standard = new Standard();
        standard.setName(name);
        standard.setDescription(description);

        // Handle use cases if provided
        if (json.has("useCaseIds")) {
            ArrayNode useCaseIds = (ArrayNode) json.get("useCaseIds");
            Set<UseCase> useCases = new HashSet<>();
            
            try {
                for (JsonNode idNode : useCaseIds) {
                    Long useCaseId = idNode.asLong();
                    UseCase useCase = jpaApi.withTransaction("default", true, (Function<EntityManager, UseCase>) entityManager ->
                        entityManager.find(UseCase.class, useCaseId)
                    );
                    if (useCase != null) {
                        useCases.add(useCase);
                    }
                }
                standard.setUseCases(useCases);
            } catch (Exception e) {
                log.warn("Error processing use case IDs: {}", e.getMessage());
            }
        }

        try {
            jpaApi.withTransaction(entityManager -> {
                entityManager.persist(standard);
            });
            log.info("Standard created successfully: {}", standard.getName());
            return created(Json.toJson(standard));
        } catch (Exception e) {
            log.error("Error creating standard: {}", e.getMessage(), e);
            return internalServerError(Json.newObject().put("error", "Could not create standard: " + e.getMessage()));
        }
    }

    public Result getStandards(Http.Request request) {
        List<Standard> standards = jpaApi.withTransaction("default", true, (Function<EntityManager, List<Standard>>) entityManager -> {
            List<Standard> result = entityManager.createQuery("SELECT s FROM Standard s ORDER BY s.name ASC", Standard.class).getResultList();
            // Force load use cases to avoid lazy loading issues
            for (Standard standard : result) {
                standard.getUseCases().size();
            }
            return result;
        });
        log.info("Retrieved {} standards.", standards.size());
        return ok(Json.toJson(standards));
    }

    public Result getStandard(Http.Request request, Long id) {
        Standard standard = jpaApi.withTransaction("default", true, (Function<EntityManager, Standard>) entityManager -> {
            Standard result = entityManager.find(Standard.class, id);
            if (result != null) {
                // Force load use cases to avoid lazy loading issues
                result.getUseCases().size();
            }
            return result;
        });
        
        if (standard == null) {
            return notFound(Json.newObject().put("error", "Standard not found."));
        }
        
        return ok(Json.toJson(standard));
    }

    public Result updateStandard(Http.Request request, Long id) {
        JsonNode json = request.body().asJson();
        if (json == null) {
            return badRequest(Json.newObject().put("error", "Expecting JSON data"));
        }

        if (id == null) {
            log.warn("Attempted to update standard with null ID");
            return badRequest(Json.newObject().put("error", "Standard ID cannot be null."));
        }

        try {
            return jpaApi.withTransaction(entityManager -> {
                Standard existingStandard = entityManager.find(Standard.class, id);
                if (existingStandard == null) {
                    return notFound(Json.newObject().put("error", "Standard not found."));
                }

                // Update name if provided
                if (json.has("name") && !json.get("name").asText().trim().isEmpty()) {
                    String newName = json.get("name").asText().trim();
                    
                    // Check if another standard with this name exists
                    List<Standard> existingByName = entityManager
                        .createQuery("SELECT s FROM Standard s WHERE LOWER(s.name) = LOWER(:name) AND s.id != :id", Standard.class)
                        .setParameter("name", newName.toLowerCase())
                        .setParameter("id", id)
                        .getResultList();
                    
                    if (!existingByName.isEmpty()) {
                        return badRequest(Json.newObject().put("error", "Another standard with this name already exists."));
                    }
                    
                    existingStandard.setName(newName);
                }

                // Update description if provided
                if (json.has("description")) {
                    existingStandard.setDescription(json.get("description").asText().trim());
                }

                // Update use cases if provided
                if (json.has("useCaseIds")) {
                    ArrayNode useCaseIds = (ArrayNode) json.get("useCaseIds");
                    Set<UseCase> useCases = new HashSet<>();
                    
                    for (JsonNode idNode : useCaseIds) {
                        Long useCaseId = idNode.asLong();
                        UseCase useCase = entityManager.find(UseCase.class, useCaseId);
                        if (useCase != null) {
                            useCases.add(useCase);
                        }
                    }
                    existingStandard.setUseCases(useCases);
                }

                entityManager.merge(existingStandard);
                // Force load use cases for response
                existingStandard.getUseCases().size();
                
                log.info("Standard updated successfully: {}", existingStandard.getName());
                return ok(Json.toJson(existingStandard));
            });
        } catch (Exception e) {
            log.error("Error updating standard with ID {}: {}", id, e.getMessage(), e);
            return internalServerError(Json.newObject().put("error", "Could not update standard: " + e.getMessage()));
        }
    }

    public Result deleteStandard(Http.Request request, Long id) {
        if (id == null) {
            log.warn("Attempted to delete standard with null ID");
            return badRequest(Json.newObject().put("error", "Standard ID cannot be null."));
        }
        Standard standard = jpaApi.withTransaction("default", true, (Function<EntityManager, Standard>) entityManager ->
            entityManager.find(Standard.class, id)
        );
        if (standard == null) {
            log.warn("Attempted to delete non-existent standard with ID: {}", id);
            return notFound(Json.newObject().put("error", "Standard not found."));
        }
        
        try {
            jpaApi.withTransaction(entityManager -> {
                entityManager.remove(standard);
            });
            log.info("Standard with ID {} deleted successfully.", id);
            return ok(Json.newObject().put("message", "Standard deleted successfully."));
        } catch (Exception e) {
            log.error("Error deleting standard with ID {}: {}", id, e.getMessage(), e);
            if (e.getCause() instanceof org.hibernate.exception.ConstraintViolationException) {
                 log.warn("Attempted to delete standard ID {} which is still in use (constraint violation).", id);
                 return status(CONFLICT, Json.newObject().put("error", "Standard is in use and cannot be deleted due to database constraints."));
            }
            return internalServerError(Json.newObject().put("error", "Could not delete standard: " + e.getMessage()));
        }
    }
}

File: src/backendng/src/main/kotlin/com/secman/controller/StandardController.kt
package com.secman.controller

import com.secman.domain.Standard
import com.secman.domain.Requirement
import com.secman.repository.StandardRepository
import com.secman.repository.UseCaseRepository
import com.secman.repository.RequirementRepository
import io.micronaut.core.annotation.Nullable
import io.micronaut.http.HttpResponse
import io.micronaut.http.HttpStatus
import io.micronaut.http.annotation.*
import io.micronaut.scheduling.TaskExecutors
import io.micronaut.scheduling.annotation.ExecuteOn
import io.micronaut.security.annotation.Secured
import io.micronaut.security.rules.SecurityRule
import io.micronaut.serde.annotation.Serdeable
import io.micronaut.transaction.annotation.Transactional
import jakarta.persistence.EntityManager
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import org.slf4j.LoggerFactory

@Controller("/api/standards")
@Secured(SecurityRule.IS_AUTHENTICATED) // All authenticated users can read
@ExecuteOn(TaskExecutors.BLOCKING)
open class StandardController(
    private val standardRepository: StandardRepository,
    private val useCaseRepository: UseCaseRepository,
    private val requirementRepository: RequirementRepository,
    private val entityManager: EntityManager
) {
    
    private val log = LoggerFactory.getLogger(StandardController::class.java)

    @Serdeable
    data class CreateStandardRequest(
        @NotBlank val name: String,
        @Nullable val description: String? = null,
        @Nullable val useCaseIds: List<Long>? = null
    )

    @Serdeable
    data class UpdateStandardRequest(
        @Nullable val name: String? = null,
        @Nullable val description: String? = null,
        @Nullable val useCaseIds: List<Long>? = null
    )

    @Serdeable
    data class ErrorResponse(
        val error: String,
        val message: String
    )

    @Get
    @Transactional(readOnly = true)
    open fun listStandards(): HttpResponse<List<Standard>> {
        return try {
            log.debug("Fetching all standards")
            
            // Query with JOIN FETCH to force loading of useCases
            val standards = entityManager.createQuery(
                "SELECT DISTINCT s FROM Standard s LEFT JOIN FETCH s.useCases ORDER BY s.name",
                Standard::class.java
            ).resultList
            
            log.debug("Found {} standards", standards.size)
            HttpResponse.ok(standards)
        } catch (e: Exception) {
            log.error("Error fetching standards", e)
            HttpResponse.serverError<List<Standard>>()
        }
    }

    @Get("/{id}")
    @Transactional(readOnly = true)
    open fun getStandard(id: Long): HttpResponse<*> {
        return try {
            log.debug("Fetching standard with id: {}", id)
            
            // Use JOIN FETCH to force loading of useCases
            val standard = entityManager.createQuery(
                "SELECT s FROM Standard s LEFT JOIN FETCH s.useCases WHERE s.id = :id",
                Standard::class.java
            ).setParameter("id", id).resultList.firstOrNull()
            
            if (standard != null) {
                log.debug("Found standard: {}", standard.name)
                HttpResponse.ok(standard)
            } else {
                log.debug("Standard not found with id: {}", id)
                HttpResponse.notFound(ErrorResponse("NOT_FOUND", "Standard not found"))
            }
        } catch (e: Exception) {
            log.error("Error fetching standard with id: {}", id, e)
            HttpResponse.serverError<Any>()
        }
    }

    @Post
    @Secured("ADMIN", "CHAMPION", "REQ") // Only privileged users can create
    @Transactional
    open fun createStandard(@Valid @Body request: CreateStandardRequest): HttpResponse<*> {
        return try {
            log.debug("Creating standard with name: {}", request.name)
            
            val trimmedName = request.name.trim()
            
            // Check for empty name after trimming
            if (trimmedName.isBlank()) {
                return HttpResponse.badRequest(ErrorResponse("VALIDATION_ERROR", "Name cannot be empty"))
            }
            
            // Check for name uniqueness (case-insensitive)
            if (standardRepository.existsByName(trimmedName)) {
                log.warn("Standard name already exists: {}", trimmedName)
                return HttpResponse.status<ErrorResponse>(HttpStatus.CONFLICT)
                    .body(ErrorResponse("CONFLICT", "Standard with this name already exists"))
            }
            
            // Create new standard
            val standard = Standard(
                name = trimmedName,
                description = request.description?.trim()
            )
            
            // Handle use case associations
            request.useCaseIds?.let { ids ->
                val useCases = ids.mapNotNull { useCaseId ->
                    useCaseRepository.findById(useCaseId).orElse(null)
                }.toMutableSet()
                standard.useCases = useCases
                log.debug("Associated {} use cases with standard", useCases.size)
            }
            
            val savedStandard = standardRepository.save(standard)
            
            // Force loading of useCases for response
            entityManager.refresh(savedStandard)
            savedStandard.useCases.size // Force lazy loading
            
            log.info("Created standard: {} with id: {}", savedStandard.name, savedStandard.id)
            HttpResponse.status<Standard>(HttpStatus.CREATED).body(savedStandard)
        } catch (e: Exception) {
            log.error("Error creating standard", e)
            HttpResponse.serverError<Any>()
        }
    }

    @Put("/{id}")
    @Secured("ADMIN", "CHAMPION", "REQ") // Only privileged users can update
    @Transactional
    open fun updateStandard(id: Long, @Valid @Body request: UpdateStandardRequest): HttpResponse<*> {
        return try {
            log.debug("Updating standard with id: {}", id)
            
            val standard = standardRepository.findById(id).orElse(null)
                ?: return HttpResponse.notFound(ErrorResponse("NOT_FOUND", "Standard not found"))
            
            // Update name if provided
            request.name?.let { newName ->
                val trimmedName = newName.trim()
                
                if (trimmedName.isBlank()) {
                    return HttpResponse.badRequest(ErrorResponse("VALIDATION_ERROR", "Name cannot be empty"))
                }
                
                // Check for name uniqueness (excluding current standard)
                val existingStandard = standardRepository.findByName(trimmedName).orElse(null)
                if (existingStandard != null && existingStandard.id != id) {
                    log.warn("Standard name already exists: {}", trimmedName)
                    return HttpResponse.status<ErrorResponse>(HttpStatus.CONFLICT)
                        .body(ErrorResponse("CONFLICT", "Standard with this name already exists"))
                }
                
                standard.name = trimmedName
            }
            
            // Update description if provided
            request.description?.let { newDescription ->
                standard.description = newDescription.trim().takeIf { it.isNotBlank() }
            }
            
            // Update use case associations if provided
            request.useCaseIds?.let { ids ->
                val useCases = ids.mapNotNull { useCaseId ->
                    useCaseRepository.findById(useCaseId).orElse(null)
                }.toMutableSet()
                standard.useCases.clear()
                standard.useCases.addAll(useCases)
                log.debug("Updated use case associations: {} use cases", useCases.size)
            }
            
            val updatedStandard = standardRepository.update(standard)
            
            // Force loading of useCases for response
            entityManager.refresh(updatedStandard)
            updatedStandard.useCases.size // Force lazy loading
            
            log.info("Updated standard: {} with id: {}", updatedStandard.name, updatedStandard.id)
            HttpResponse.ok(updatedStandard)
        } catch (e: Exception) {
            log.error("Error updating standard with id: {}", id, e)
            HttpResponse.serverError<Any>()
        }
    }

    @Delete("/{id}")
    @Secured("ADMIN", "CHAMPION", "REQ") // Only privileged users can delete
    @Transactional
    open fun deleteStandard(id: Long): HttpResponse<*> {
        return try {
            log.debug("Deleting standard with id: {}", id)
            
            val standard = standardRepository.findById(id).orElse(null)
                ?: return HttpResponse.notFound(ErrorResponse("NOT_FOUND", "Standard not found"))
            
            standardRepository.delete(standard)
            
            log.info("Deleted standard: {} with id: {}", standard.name, id)
            HttpResponse.ok(mapOf("message" to "Standard deleted successfully"))
        } catch (e: Exception) {
            log.error("Error deleting standard with id: {}", id, e)
            
            // Check if it's a constraint violation (standard in use)
            if (e.message?.contains("constraint", ignoreCase = true) == true ||
                e.message?.contains("foreign key", ignoreCase = true) == true) {
                return HttpResponse.status<ErrorResponse>(HttpStatus.CONFLICT)
                    .body(ErrorResponse("CONFLICT", "Cannot delete standard - it may be in use"))
            }
            
            HttpResponse.serverError<Any>()
        }
    }

    @Get("/{id}/requirements")
    @Transactional(readOnly = true)
    open fun getStandardRequirements(id: Long): HttpResponse<*> {
        return try {
            log.debug("Fetching requirements for standard with id: {}", id)
            
            // First get the standard with its use cases
            val standard = entityManager.createQuery(
                "SELECT s FROM Standard s LEFT JOIN FETCH s.useCases WHERE s.id = :id",
                Standard::class.java
            ).setParameter("id", id).resultList.firstOrNull()
            
            if (standard == null) {
                log.debug("Standard not found with id: {}", id)
                return HttpResponse.notFound(ErrorResponse("NOT_FOUND", "Standard not found"))
            }
            
            // Get all requirements for the use cases associated with this standard
            val requirements = mutableListOf<Requirement>()
            standard.useCases.forEach { useCase ->
                val useCaseRequirements = requirementRepository.findByUsecaseId(useCase.id!!)
                requirements.addAll(useCaseRequirements)
            }
            
            // Remove duplicates (in case a requirement is associated with multiple use cases)
            val uniqueRequirements = requirements.distinctBy { it.id }
            
            log.debug("Found {} requirements for standard: {}", uniqueRequirements.size, standard.name)
            HttpResponse.ok(uniqueRequirements)
        } catch (e: Exception) {
            log.error("Error fetching requirements for standard with id: {}", id, e)
            HttpResponse.serverError<Any>()
        }
    }
}